src/migrations/rebase.rs

```diff
@@ -0,0 +1,286 @@
+        }
+
+        let expected_id = migration_file.data.expected_id(&migration_text)?;
+
+        if id != expected_id {
+            migration_text = migration_file.data.replace_id(&migration_text, &expected_id);
+            migration_file.data.id = expected_id.clone();
+            changed_ids.insert(id, expected_id);
+        }
+
+        fs::write(&migration_file.path, migration_text)?;
+    }
+
+    Ok(())
+}
+
+pub async fn do_rebase(connection: &mut Connection, context: &Context, rebase_migrations: RebaseMigrations) -> anyhow::Result<()> {
+    let temp_dir = tempfile::tempdir()?;
+    let temp_ctx = Context {
+        schema_dir: temp_dir.path().to_path_buf(),
+        edgedb_version: None,
+        quiet: false,
+    };
+
+    // write all the migrations to disc.
+    let flattened_migrations = rebase_migrations.flatten()?;
+    for migration in &flattened_migrations {
+        create::write_migration(&temp_ctx, migration, false).await?;
+    }
+
+    // update the IDs of the migrations for the rebase, since we're changing the history the IDs can be invalid so we
+    // need to update them.
+    rebase_migration_ids(&temp_ctx).await?;
+
+    // remove the old migrations
+    for old in migration::read_names(context).await? {
+        fs::remove_file(old)?;
+    }
+
+    // move the new migrations from temp to schema dir
+    let mut last: Option<&RebaseMigrationKind> = None;
+    for (index, from) in migration::read_names(&temp_ctx).await?.iter().enumerate() {
+        let rebase_migration = flattened_migrations.get(index);
+        let to = context
+            .schema_dir
+            .join("migrations")
+            .join(from.file_name().expect(""));
+
+        if let Some(migration) = rebase_migration {
+            if last != Some(&migration.kind) {
+                match migration.kind {
+                    RebaseMigrationKind::Target => {
+                        eprintln!("\nNew migrations on target branch:")
+                    },
+                    RebaseMigrationKind::Source => {
+                        eprintln!("\nMigrations to rebase:")
+                    },
+                    RebaseMigrationKind::Base => {}
+                }
+            }
+
+            match migration.kind {
+                RebaseMigrationKind::Target | RebaseMigrationKind::Source => {
+                    print::success_msg("Writing", to.display());
+                }
+                RebaseMigrationKind::Base => {}
+            }
+        }
+
+
+        fs::copy(from, to)?;
+
+        last = rebase_migration.map(|v| &v.kind);
+    }
+
+    // apply the new migrations
+    let migrations: IndexMap<String, MigrationFile> = migration::read_all(context, true).await?.into_iter()
+        .filter(|(id, _)| rebase_migrations.target_migrations.contains_key(id))
+        .collect();
+
+    migrate::apply_migrations(connection, &migrations, context, true).await?;
```

Comment:
Just a suggestion: it would make sense to have this chunk moved into `branch::rebase::rebase` function and rename this function to something like `write_rebased_migration_files`.

Reply:
That requires `migrate` and `migration` become public modules, is that okay?

Reply:
Oh nvm, I see what you mean 💯